datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -841,22 +838,50 @@ public void setDatastreamTaskStateForKey(DatastreamTask datastreamTask, String k
   }
 
   /**
-   * Remove instance assignment nodes whose instances are dead
+   * Remove instance assignment nodes whose instances are dead. This also
+   * removes the node from under /connector if present in unusedTasks.
    *
    * NOTE: this should only be called after the valid tasks have been
    * reassigned or become safe to discard per strategy requirement.
    * Coordinator is expected to cache the "current" assignment before
    * invoking the assignment strategy and pass the saved assignment
    * to us to figure out the obsolete tasks.
    * @param instances Instances for which nodes are to be cleaned up
+   * @param unusedTasks Unused tasks that can be cleaned up once they
+   *                    are removed from dead instances.
    */
-  public void cleanupDeadInstanceAssignments(List<String> instances) {
+  public void cleanupDeadInstanceAssignmentsAndUnusedTasks(List<String> instances,
+      Set<DatastreamTask> unusedTasks) {
     List<String> deadInstances = getAllInstances();
+    Map<String, DatastreamTask> unusedTasksByName = new HashMap<>();
+    unusedTasks.forEach(unusedTask -> unusedTasksByName.put(unusedTask.getDatastreamTaskName(), unusedTask));
     deadInstances.removeAll(instances);
+
+    LOG.debug("unusedTasks before cleanup: {}", unusedTasksByName.keySet());
     if (deadInstances.size() > 0) {
       LOG.info("Cleaning up assignments for dead instances: " + deadInstances);
 
       for (String instance : deadInstances) {
+        // get the old assignment from ZooKeeper
+        Set<String> oldAssignments = new HashSet<>(getInstanceAssignment(instance));
+
+        oldAssignments.forEach(oldAssignment -> {
+          removeTaskNodes(instance, oldAssignment);
+          // /<cluster>/connectors/<connectorType>/<task> node should always be deleted after
+          // /<cluster>/instances/<task>/.
+          // If a node is present under /instance, the code assumes that the node will be present under
+          // /<cluster>/connectors/<connectorType>/ as well. So, the cleanup order is very important.
+
+          // If the thread gets interrupted after removing node under /instance, but before deleting node
+          // under /<cluster>/connectors/<connectorType>/, next leader will take care of cleaning this orphan node in
+          // cleanUpOrphanConnectorTask call.
```

Comment:
I like this comment a lot but:
- It doesn't say that `removeTaskNodes()` removes `/<cluster>/instances/<task>/`
- The path isn't exactly `/<cluster>/instances/<task>/`
- It assumes a bit of background

How about we write this as:
```java
        oldAssignments.forEach(oldAssignment -> {
          // We have two znodes to clean up for every datastream task
          //  1. /<cluster>/instances/{instanceId}/assignments/<task>
          //  2. /<cluster>/connectors/<connectorType>/<task>
          // 
          // (1) should always be deleted first. If we were to delete (2) first,
          // and the code got interrupted (e.g. due to shutdown), we would run
          // the risk of having the next leader assign a task with no presence
          // in (2) to a live host, which may cause errors in other pieces of
          // code because datastream task entries in (2) are needed for host-
          // independent operations like locking tasks and check-pointing progress.
          //
          // On the other hand, when we delete (1) before (2), if the code is 
          // interrupted in between the two operations, the next leader will not
          // assign that task to any host; only (2) will become an orphan node 
          // and the next leader will take care of deleting it in cleanUpOrphanConnectorTask().
          
          // Delete (1)
          removeTaskNodes(instance, oldAssignment);
          
          DatastreamTask unusedTask = unusedTasksByName.remove(oldAssignment);
          if (unusedTask != null) {
            // Delete (2)
            deleteConnectorTask(unusedTask.getConnectorType(), unusedTask.getDatastreamTaskName());
            unusedTasks.remove(unusedTask);
          }
        });
```